Digital normalization with BBnorm

erikrikarddaniel opened the pull request:
<!--
# nf-core/mag pull request

Many thanks for contributing to nf-core/mag!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

I added BBnorm, run with `--bbnorm` (with two more params), which will normalize sequencing depth by reducing the number of reads for overrepresented kmers as well as removing very rare kmers. 

I've only run the tool on the small test cases, and can't say if and when it's useful to use on full-sized data. I hence didn't update `usage.md`.

I didn't add the new test case (tests the combination of `--bbnorm` and `--coassemble_group`.

The tool is not published in a journal, and they just want people to refer to the Sourceforge page. I didn't know how to do this in `CITATIONS.md` but provide an attempt.

jfy133 commented on CITATIONS.md:
```diff
@@ -14,6 +14,10 @@
   > Schubert, M., Lindgreen, S., and Orlando, L. 2016. "AdapterRemoval v2: Rapid Adapter Trimming, Identification, and Read Merging." BMC Research Notes 9 (February): 88. doi: 10.1186/s13104-016-1900-2
 
+- [BBnorm/BBTools](http://sourceforge.net/projects/bbmap/)
+
+  > Only citable by the Sourceforge URL.
```
```suggestion
```

jfy133 commented on conf/test_bbnorm.config:
```diff
@@ -0,0 +1,35 @@
+    busco_clean                   = true
+    gtdb                          = false
+    skip_concoct                  = true
+    bbnorm                        = true
```
Is an entire test config necessary? I think you can include in a basic test, unless it takes a long time

erikrikarddaniel replied:
Do you mean add it to e.g. `test.config`?

d4straub replied:
Yes I think he did.
If this test config is indeed needed, I'd remove as many steps as possible that are tested in other configs, to have as minimal overlap as possible while still testing as many toll combinations as possible.

erikrikarddaniel replied:
I found that there was no test for the grouped assembly, so I kept this and deleted many other steps. Runs very quickly now.

jfy133 commented on workflows/mag.nf:
```diff
@@ -445,26 +478,36 @@ workflow MAG {
+                )
+                POOL_SHORT_SINGLE_READS.out.reads
+                    .mix(POOL_PAIRED_READS.out.reads)
+                    .set { ch_short_reads_spades }
```
```suggestion
```

erikrikarddaniel commented:
> * Missing output docs
> 
I actually forgot to organize what to save in the output by default. And, now that I'm reminded, I wouldn't mind some recommendations from other contributors to the pipeline: Do you think it makes sense to save any of the output from either of the two new modules? Can be quite large files.

>      * Missing SeqTK citation/documentation(?)
> 
Thanks. I had forgotten I included this tool as it took so many days from implementing the code to actually do the PR.

> 
> To be clear, `seqtk merge` doesn't do paired-end merging/collapsing right? It just referse to merging the two files so all the reads are interleaved? It would be good to clarify in the comment around that module, it confused me at first and I wondered why a different read-collapser was being used.
> 
It's interleaving. We're using it in metatdenovo, but didn't write the module. You want it better documented in the module repo, I suppose. I'll try to remember that.

> Note I can't really comment on the validity of method etc (but logically the code makes sense to me)

Good. I'm not sure either, but there's one clear use case: When you can't, but really want to, get a coassembly through.

erikrikarddaniel commented:
> > I actually forgot to organize what to save in the output by default. And, now that I'm reminded, I wouldn't mind some recommendations from other contributors to the pipeline: Do you think it makes sense to save any of the output from either of the two new modules? Can be quite large files.
> 
> Good question. Interleaving probably not. I didn't actually look what digital normalization does either so I can't say if the output would be useful grimacing . I'll leave that to others to decide :)

OK, I kept it and updated `output.md`.
> 
> > Thanks. I had forgotten I included this tool as it took so many days from implementing the code to actually do the PR.
> 
> > It's interleaving. We're using it in metatdenovo, but didn't write the module. You want it better documented in the module repo, I suppose. I'll try to remember that.
> 
> It could do, but I actually meant more here in the pipeline code. I have never used interleaving data myself and 'merge' always means 'collapse' to me personally., so I don't want to come back to this section of the pipeline and get confused in the future grimacing .

Good point, updated.

d4straub commented on conf/modules.config:
```diff
@@ -136,6 +136,25 @@ process {
+        publishDir = [
+            [
+                path : { "${params.outdir}/bbmap/bbnorm/logs" },
+                mode : 'copy',
```
I think that would be better:
```suggestion
                mode : params.publish_dir_mode,
```
also below in the second statement

d4straub commented on docs/output.md:
```diff
@@ -129,6 +129,18 @@ NanoPlot is used to calculate various metrics and plots about the quality and le
+
+If the pipeline is called with the `--bbnorm` option, it will normalize sequencing depth by removing reads to 1) reduce coverage of very abundant kmers and 2) delete very rare kmers (see `--bbnorm_target` and `--bbnorm_min` parameters).
+When called in conjunction with `--coassemble_group`, BBnorm will operate on interleaved (merged) FastQ files, producing only a single output file.
+The resulting FastQ files are save together with log output.
```
Uff, I'm not sure that is a good idea, that is potentially large? Maybe make the saving optional, e.g. similar to https://nf-co.re/mag/2.3.0/parameters#save_clipped_reads and use it as in https://github.com/nf-core/mag/blob/c9468cb9154402c4a4cab78664ee50c24a91f6c3/conf/modules.config#L50

erikrikarddaniel replied:
Good that you reminded me (I should have remembered this strategy from our discussion) -- done!

d4straub commented on modules/local/spades.nf:
```diff
@@ -20,14 +20,15 @@ process SPADES {
     script:
     def args = task.ext.args ?: ''
     maxmem = task.memory.toGiga()
+    def readstr = meta.single_end ? "--12 ${reads}" : "-1 ${reads[0]} -2 ${reads[1]}"
```
```suggestion
    def readstr = meta.single_end ? "--12 ${reads}" : "-pe1-1 ${reads[0]} -pe1-2 ${reads[1]}"
```
Or is there a reason not to use "pe"? (I might be not up to date with SPAdes usage)

d4straub reviewed (commented):
Just a few comments, looks good so far I think. But I also do not have experience with bbnorm / digital normalization.

erikrikarddaniel commented:
I think I've done the things you asked for @jfy133 and @d4straub.

d4straub commented:
Sorry, I had an extended eastern vacation.

According to https://github.com/ablab/spades#specifying-multiple-libraries I made a mistake in the above code suggestion: instead of `-pe1-1`/`-pe1-2`, `--pe1-1`/`--pe1-2` seems correct. As far as I see only one library should arrive at that point, so `-1` / `-2` should be fine as well I assume.

It might be good to add in `modules/local/spades.nf` an explanation/comment, or is there a typo:
`    def readstr = meta.single_end ? "--12 ${reads}" : "-1 ${reads[0]} -2 ${reads[1]}"`
seems to expect a file with interlaced forward and reverse paired-end reads, not single-end data (which would be `-s`?!).

I am fine with the PR, but @jfy133 requested changes and has to give his ok to the changes.

erikrikarddaniel commented:
> Sorry, I had an extended eastern vacation.
> 
Well deserved! :-)

> According to https://github.com/ablab/spades#specifying-multiple-libraries I made a mistake in the above code suggestion: instead of `-pe1-1`/`-pe1-2`, `--pe1-1`/`--pe1-2` seems correct. As far as I see only one library should arrive at that point, so `-1` / `-2` should be fine as well I assume.
> 
The `--pe1-1` and `-1` syntax is a bit confusing at the moment, but I chose to keep it the old style, i.e. `-1`.

> It might be good to add in `modules/local/spades.nf` an explanation/comment, or is there a typo: ` def readstr = meta.single_end ? "--12 ${reads}" : "-1 ${reads[0]} -2 ${reads[1]}"` `--12` seems to expect a file with interlaced forward and reverse paired-end reads, not single-end data (which would be `-s`?!).
> 
It turns out `-s` is not supported by metaspades yet, although output as part of the help message. I've added a comment. An alternative would be to add a different `meta` variable, e.g. `meta.interleaved`. The drawback with that is that the nf-core module doesn't support that. OTOH, the nf-core module supports quite intricate input of combinations of read files to spades via a yaml file.

> I am fine with the PR, but @jfy133 requested changes and has to give his ok to the changes.

jfy133 commented on docs/output.md:
```diff
@@ -129,6 +129,18 @@ NanoPlot is used to calculate various metrics and plots about the quality and le
 
+## Digital normalization with BBnorm
+
+If the pipeline is called with the `--bbnorm` option, it will normalize sequencing depth by removing reads to 1) reduce coverage of very abundant kmers and 2) delete very rare kmers (see `--bbnorm_target` and `--bbnorm_min` parameters).
```
```suggestion
If the pipeline is called with the `--bbnorm` option, it will normalize sequencing depth of libraries prior assembly by removing reads to 1) reduce coverage of very abundant kmers and 2) delete very rare kmers (see `--bbnorm_target` and `--bbnorm_min` parameters).
```

jfy133 commented on docs/output.md:
```diff
@@ -129,6 +129,18 @@ NanoPlot is used to calculate various metrics and plots about the quality and le
+
+- `bbmap/bbnorm/[sample]\*.fastq.gz
+- `bbmap/bbnorm/log/[sample].bbnorm.log
+
```
Missing closing tags?

jfy133 replied:
You missed the closing rtags @erikrikarddaniel 😱 dev docs are broken! I'll make a PR now (skipping ci) form you to approve

jfy133 commented on workflows/mag.nf:
```diff
@@ -309,6 +311,31 @@ workflow MAG {
+                .mix(ch_short_reads.filter { it[0].single_end })
+                .map { [ [ id: sprintf("group%s", it[0].group), group: it[0].group, single_end: true ], it[1] ] }
+                .groupTuple()
+                .set { ch_bbnorm }
```
```suggestion
```

jfy133 commented on workflows/mag.nf:
```diff
@@ -309,6 +311,31 @@ workflow MAG {
+            )
+            ch_versions = ch_versions.mix(SEQTK_MERGEPE.out.versions.first())
+            // Combine the interleaved pairs with any single end libraries. Set the meta.single_end to true (used by the bbnorm module).
+            SEQTK_MERGEPE.out.reads
```
```suggestion
             ch_bbnorm = SEQTK_MERGEPE.out.reads
```

jfy133 reviewed (approved):
Sorry for the delay, I made the 'mistake' of taking a holiday and had a huge namount of backlog to catch up with...

Last few changes and I think you're OK, as  @d4straub is happy with the 'scientific' content'.

erikrikarddaniel commented:
Thanks @jfy133: fantastic attention to detail! :-)

erikrikarddaniel commented:
I did add them, something must have happened...

jfy133 commented:
:scream:
